File: Java/src/main/java/biz/shop/Rabattrechner.java
package biz.shop;

import java.math.BigDecimal;

public class Rabattrechner {

    private final BigDecimal noRabatt = BigDecimal.ZERO;
    private final BigDecimal threshold = new BigDecimal(1000);
    private final BigDecimal percentage = new BigDecimal(3).divide(new BigDecimal(100));

    public BigDecimal calcRabattFor(BigDecimal purchase) {
        if (purchase.compareTo(threshold) >= 0) {
            return purchase.multiply(percentage);
        } else {
            return noRabatt;
        }
    }

}

File: PHP/src/RabattRechner.php
<?php
namespace Shop;

use Brick\Math\BigDecimal;

class RabattRechner
{

    private $noDiscount;
    private $levels;

    public function __construct()
    {
        $this->noDiscount = BigDecimal::zero();
        $this->levels = [];
    }

    public function addDiscountLevel(BigDecimal $threshold, $percentage)
    {
        $decimalPercentage = BigDecimal::of($percentage)
            ->exactlyDividedBy(BigDecimal::of(100));
        array_push($this->levels, new DiscountLevel($threshold, $decimalPercentage));
    }

    public function discountFor(BigDecimal $purchase)
    {
        foreach ($this->levels as $level) {
            if ($purchase->compareTo($level->threshold) >= 0) {
                return $purchase->multipliedBy($level->percentage);
            }
        }
        return $this->noDiscount;
    }

}
